### ISSUE TITLE
[Fleet]: when upgrading an agent it stays 'unhealthy' and doesn't stream new data for System integration

### ISSUE THREAD

**Issue by amolnater-qasource** [elastic/beats#26177]
**Kibana version:** 7.14.0 Snapshot Kibana cloud environment

**Host OS and Browser version:** Windows 10 x64, All

**Build Details:**
```
  Artifact link used: https://artifacts.elastic.co/downloads/beats/elastic-agent/elastic-agent-7.13.1-windows-x86_64.zip
  Build: 41498
  Commit: 5cab87e2069dd31848490f83964291fc802d6889
```

**Preconditions:**
1. 7.14.0 Snapshot Kibana cloud environment should be available.
2. 7.13.1(released) Windows 10 x64 agent must be installed with policy having System and Endpoint Security integration._[We have used non-default namespace for better tracking of data streams.]_ 

**Steps to reproduce:**
1. Login to Kibana environment.
2. Trigger Upgrade button for 7.13.1 agent under Agents tab.
3. After upgrade observe Agent "Unhealthy" status.
4. Navigate to Data Streams tab and observe no new data for System integration.

**Expected Result:**
7.13.1 Agent should be upgraded to 7.14 without any errors and new data for all integrations should generate under Data Streams tab.

**Logs:**
[logs_elastic-agent-2d80f6.zip](https://github.com/elastic/kibana/files/6608739/logs_elastic-agent-2d80f6.zip)
[logs_elastic-agent-71f8f0.zip](https://github.com/elastic/kibana/files/6608752/logs_elastic-agent-71f8f0.zip)

**Note:**
- Filebeat, Metricbeat and Endpoint Security services are in RUNNING state.

**Screenshots:**
![Agent status](https://user-images.githubusercontent.com/77374876/121015802-06174e80-c7b9-11eb-9fe6-9407abbed30e.jpg)

**Comment by EricDavisX** [comment]
I closed an upgrade related ticket i had logged a month ago, https://github.com/elastic/beats/issues/25668 - this one has much more helpful logging info in it.  Let's track the concern here.  

It needs to be re-tested after we get a coming 7.14 builds with the Metricbeat re-start / crash loop fix (from here:  https://github.com/elastic/beats/issues/26041 )

**Comment by amolnater-qasource** [comment]
Hi @EricDavisX 
Thanks for the feedback, we will re-test this when Metricbeat re-start/crash loop issue will be fixed.

Thanks
QAS

**Comment by amolnater-qasource** [comment]
Hi @EricDavisX 
We have revalidated this issue on 7.14.0 Snapshot Kibana cloud environment and found it **fixed** now.
We have upgraded 7.13.2 Snapshot agent to 7.14.0 Snapshot.

**Observations after upgrade:**
- Agent remain Healthy after upgrade.
- New data for System Integration was generating under Data Streams tab.

**Build details:**
```
Build: 41559
Commit: 9838db392e7fcfc12f004b68fb1b09739f131148
Artifact Link: https://snapshots.elastic.co/7.13.2-f55bee50/downloads/beats/elastic-agent/elastic-agent-7.13.2-SNAPSHOT-windows-x86_64.zip
```
 
Hence, we are closing this ticket.

Thanks
QAS

**Comment by amolnater-qasource** [comment]
Hi @EricDavisX 
We have revalidated this by upgrading 7.13.2 released agent on 7.14.0 Snapshot Kibana and observed that agent goes "Unhealthy".

However we haven't observed this issue for Snapshot agent(as shared under https://github.com/elastic/beats/issues/26177#issuecomment-861436077) for which we have closed this issue.
Could you please confirm if we don't expect released agent to upgrade to snapshot agent?

We are re-opening this ticket until further clarity.

Thanks
QAS

**Comment by EricDavisX** [comment]
This absolutely is a bug, the Agent upgrade is fairly simple in this regards - it looks at the Kibana version and uses the existence or lack -SNAPSHOT in the name to determine what to upgrade to, nothing else is involved.

Having said that - other bugs may be logged for the problem, this may be just one of many scenarios where a 'restart' usage of Agent isn't working.  If the bug exists in older version of Agent it may be impacting the upgrade.  I can't say for sure, so I will add it to the 'urgent' issues list and tag @michalpristas who knows the upgrade (and recent related bug fixes for Agent restart fairly well).

Please do re-test and attach current logs, because we have been fixing these bugs for a while and new / fresh logs will help make fastest use of Engineering time when fixing it.  Thank you. @amolnater-qasource

**Comment by amolnater-qasource** [comment]
Hi @EricDavisX 
We have revalidated this issue and we are still able to reproduce this on 7.14.0 Snapshot Kibana cloud environment.
Please find below the logs attached for the same.

**Logs:**
[logs-elastic-agent-1fd6b2.zip](https://github.com/elastic/beats/files/6724130/logs-elastic-agent-1fd6b2.zip)
[logs-elastic-agent-686ba4.zip](https://github.com/elastic/beats/files/6724131/logs-elastic-agent-686ba4.zip)

Further we observed that Endpoint Security version is not upgraded after agent upgrade.
**Screenshot:**
![3](https://user-images.githubusercontent.com/77374876/123595259-53ae3680-d80e-11eb-9d63-85130e8fda56.JPG)

Please let us know if anything else is required from our end.
Thanks
QAS

**Comment by michalpristas** [comment]
@amolnater-qasource could you retest with 7.13.3 or 7.14 BC,
this is most likely related to #26176

**Comment by amolnater-qasource** [comment]
Hi @michalpristas 
We have revalidated this issue on 7.14 BC cloud-qa Kibana environment.
Could you please confirm that do BC builds support Agent upgrade?
- As even on earlier version BC builds we have failed to cover Agent upgrade tests.

We have covered following Scenarios:
**Scenario 01:**
We attempted to upgrade **7.13.3 BC-2** agent upgrade on **7.14.0 BC** environment.
- Agent didn't upgrade and went to unhealthy state.

**Screenshot:**
![12](https://user-images.githubusercontent.com/77374876/124888549-d2b42380-dff3-11eb-87eb-92f49219e47c.JPG)

**Logs for BC build:**
[BC_build_logs.zip](https://github.com/elastic/beats/files/6782892/BC_build_logs.zip)

**Scenario 02:**
We attempted to upgrade **7.13.3 (1)BC-2 and (2)Released agent** upgrade on **7.14.0 Snapshot environment**.
- Agent upgraded to 7.14.0 and went to unhealthy state.

**Screenshot:**
![13](https://user-images.githubusercontent.com/77374876/124888998-42c2a980-dff4-11eb-98fa-a7672e6d197f.JPG)

**Logs for Snapshot build with Released 7.13.3 agent:**
[elastic-agent-3ddad4_logs.zip](https://github.com/elastic/beats/files/6782938/elastic-agent-3ddad4_logs.zip)
[elastic-agent-731003_logs.zip](https://github.com/elastic/beats/files/6782941/elastic-agent-731003_logs.zip)

Further we observed that Endpoint Security is not upgraded on 7.14.0 Snapshot build.
![14](https://user-images.githubusercontent.com/77374876/124890130-5b7f8f00-dff5-11eb-863b-8d22bd457068.JPG)

Please let us know if anything else is required from my end.

Thanks
QAS

**Comment by michalpristas** [comment]
scenario 1 is expected, BC does not have SNAPSHOT suffix so it tries to update agent to 7.14 and download it from official repo. this is not possible as 7.14 was not published yet

scenario 2 i need to verify but this looks like it has issues with endpoint `operation 'Exec' failed (return code: 1): (command had no output): operation 'Exec' failed (return code: 1): (command had no output)`

**Comment by michalpristas** [comment]
can this be retested using 7.13.3 snapshot to 7.14.0 snapshot. 
and 7.13.3 (released) to 7.14.0 snapshot
i tried multiple times on cloud qa and failed to reproduce. 
<img width="971" alt="image" src="https://user-images.githubusercontent.com/1522652/124932116-16faf000-e003-11eb-8272-040a99037283.png">

<img width="1049" alt="image" src="https://user-images.githubusercontent.com/1522652/124931693-c1264800-e002-11eb-8029-dcd5958449c7.png">

i can retest in your env maybe it's an env thing

**Comment by amolnater-qasource** [comment]
Hi @michalpristas 
We have validated this issue by upgrading released agent of 7.13.3 and its still **reproducible**.

**Build details:**
```
Build: 42405
Commit: d449412c8ca14a87a10527ffdc3967d42c12f34a
```

**Screenshot:**
![6](https://user-images.githubusercontent.com/77374876/125056613-06f81480-e0c6-11eb-90be-2ed31bc70dec.JPG)

**Logs:**
[endpoint-000000.zip](https://github.com/elastic/beats/files/6790297/endpoint-000000.zip)
[elastic-agent-3ddad4_logs.zip](https://github.com/elastic/beats/files/6790300/elastic-agent-3ddad4_logs.zip)
[elastic-agent-ceb338_logs.zip](https://github.com/elastic/beats/files/6790303/elastic-agent-ceb338_logs.zip)

Please let us know if anything else is required from my end.

Thanks
QAS

**Comment by michalpristas** [comment]
i will try this on amol's env, seems like one of these issues which appears only on qas vms :D

**Comment by michalpristas** [comment]
on your environment i see this executed from 7.14 install directory of agent

```
 .\endpoint-security.exe verify                                                                                                                                                                                                     Program 'endpoint-security.exe' failed to run: Windows cannot verify the digital signature for this file. A recent hardware or software change might have installed a file that is signed incorrectly or damaged, or that might be malicious software from an unknown sourceAt line:1 char:1
+ .\endpoint-security.exe verify
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
At line:1 char:1
+ .\endpoint-security.exe verify
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ResourceUnavailable: (:) [], ApplicationFailedException
    + FullyQualifiedErrorId : NativeCommandFailed
```

when running it from 7.13.3 repository it is ok.  and does not fail.

so i guess agent does the correct thing and endpoint archive is broken/not signed ... 

@EricDavisX do you recall what's wrong with the archive in case of this error? i recall there was some option on OS level to skip verification but i'm not sure if i recall correctly

**Comment by EricDavisX** [comment]
hi. i don't recall Endpoint lore very well, I have pinged in slack and will tag @ferullo if he can help confirm this scenario 02 usage of Endpoint Security snapshot 7.14 build 'verify' call

**Comment by EricDavisX** [comment]
I spoke with Ferullo in slack and he commented, foremost, when seeing the above 'verify' error:
Daniel Ferullo  9:50 AM
Windows cannot verify the digital signature for this file you are trying to run an unsigned endpoint (i.e. not a BC snapshot) on a VM with with test signing disabled.

A little more detail about the command:
verify checks that the installed endpoint (c:\Program Files\Elastic\Endpoint\elastic-endpoint.exe) reports the same output from the version  as the one that verify is run for. Does that make sense?

9:45
if verify fails [you can] re-run the endpoint with version and then see what the installed one outputs for version as well. 

From Eric:
So, In this way it is a specific tool for us to 'verify' the install of Endpoint'

@amolnater-qasource @michalpristas this is the best news we could have!  It is likely a configuration problem and not a product bug.  Amol, please do re-test (and I'll run one as well and we can close this out when confirmed.  

From the Engineering side, I admit there are a lot of manual steps to set up correctly here, it may be worthwhile to put the windows host automation of Agent / Endpoint upgrade into higher priority now that the Windows environment is (or will be imminently) available for us to automate against.

**Comment by amolnater-qasource** [comment]
Hi @EricDavisX 
We were earlier installing released Agent on Test Signing OFF machine, which was expected.
However today we have attempted to install 7.13.3 agent with Test Signing ON machine and we are able to upgrade released agent successfully to 7.14.0 Snapshot.

- Agent remained Healthy throughout.

**Screenshot:**
![5](https://user-images.githubusercontent.com/77374876/125604197-901e0396-4943-42b2-8b02-47bdf961ba57.JPG)

**Endpoint verify and version command:**
![6](https://user-images.githubusercontent.com/77374876/125604407-229fd609-2abc-471c-8173-5546613579c3.JPG)

Hence we are closing this issue.
Please let us know if anything else is required from our end.

Thanks
QAS
